GH-477 WIP: Merge N4IDL as a language variant of N4JS

lbeurerkellner opened the pull request:
Issue: https://github.com/eclipse/n4js/issues/477

This PR is the current version of introducing N4IDL as N4JS language variant.

As many things aren't fully decided yet, this PR is marked as WIP. However, as discusses with @jpilgrim, the goal of this PR is not to fix all remaining issues with N4IDL but rather to be a first working version of it. 

In the following I will outline the main-steps that were required in order to get a first running version.

1. Moved N4IDL.xcore in parts to N4JS.xcore and in parts to TypeRefs.xcore
2. Merge N4IDL.xtext into N4JS.xtext and TypeExpressions.xtext
3. Add support for versions in the N4JS TypesBuilder
4. Integrate N4IDLVersionAwareScopeProvider with the N4JSScopeProvider
5. Merge N4JSJavaScriptVariantHelper 
6. Adapt imports validation and unused import analysis to be aware of versions
7. Moved test bundles and removed obsolete com.enfore.n4idl bundles
8. Migrate N4IDL transpiler from IDE-2842 branch (make N4IDL executable)
9. Disallow top-level statements in N4IDL
10. Remove 'migration' keyword
> Migrations can be declared as follows:
> ```
> @Migration
> function migrateA(a1: A#1) : A#2 { (...) }
> ```
11. Disallow versioned types + migrations in N4JS
12. Introduce new token VERSION in order to disambiguate grammar
> This was required, as otherwise a property access on static fields of versioned types could not be parsed correctly. For instance, the parser (lexer) would recognise the input `T#1.staticField` as a sequence of `ID("T") ID("#") DECIMAL("1.")` *Lexer error* at "s"

TODO:

* After a discussion with @kduske-n4 today, we came to the conclusion that supporting versioned import specifiers (e.g. `import {A#1} from "A"`) may cause issues in N4IDL files. Alternatively, we could instead interpret regular import statements, such as `import {A} from "A"`, as an import-all-versions statement. When transpiling N4IDL, unused versions can still automatically be sanetised. I am currently investigating how such an import semantic could be implemented.

mor-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/organize/imports/ImportProvidedElement.java:
```diff
@@ -25,10 +26,11 @@
 
 	/** name as used in script (can be alias) */
-	public String localname;
+	private final String localname;
```
nice clean up! but please do the same clean up for fields `importSpec`, `tmodule`, `used` in same class.

mor-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/resource/N4JSResourceDescriptionStrategy.java:
```diff
@@ -118,17 +118,57 @@ public boolean createEObjectDescriptions(final EObject eObject, IAcceptor<IEObje
+	 * @returns An immutable user-data map
+	 */
+	protected Map<String, String> createClassUserData(final Map<String, String> immutableUserData,
+			final TClass tClass) {
```
1. this was factored out from method `#internalCreateEObjectDescription()`, so I'd place this new method directly below `#internalCreateEObjectDescription()`.
2. why protected? I can't find anyone who overrides this atm. Would private make sense?

mor-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/UsageAwareObjectDescription.java:
```diff
@@ -0,0 +1,80 @@
+ * An {@link IEObjectDescription} that has the callback-method {@link #markAsUsed()} that is invoked when this object
+ * description is used to bind a name.
+ */
+public abstract class UsageAwareObjectDescription<T extends IEObjectDescription> implements IEObjectDescription {
```
1. the "E" in front of "Object" is missing in the name, here.
2. Interface `UsageAwareEObjectDescription` should then be renamed to `IUsageAwareEObjectDescription` (note the "I" at the beginning).

mor-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/N4IDLVersionResolver.xtend:
```diff
@@ -0,0 +1,157 @@
+ * version. Generally, when resolving an actual version of a versioned type, this implementation will delegate to
+ * {@link VersionHelper#findClassifierWithVersion(TClassifier, int)}.
+ */
+class N4IDLVersionResolver implements VersionResolver {
```
maybe we should now move this class and the implemented interface to package `org.eclipse.n4js.n4idl.versioning`. What do you think?

mor-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSImportValidator.xtend:
```diff
@@ -40,7 +40,8 @@ import static org.eclipse.n4js.validation.IssueCodes.*
 
-/** Validations for the import statements. */
+/**
+ */
```
why remove this comment?

mor-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSImportValidator.xtend:
```diff
@@ -146,17 +147,20 @@ class N4JSImportValidator extends AbstractN4JSDeclarativeValidator {
 	 * @return name from NamedImportSpecifier or AST-text if unresolved.
 	 */
-	private def String importedElementErrorName(NamedImportSpecifier spec) {
+	protected def String computeImportSpecifierName(NamedImportSpecifier spec) {
```
why increase visibility, here?

mor-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSImportValidator.xtend:
```diff
@@ -339,19 +349,27 @@ class N4JSImportValidator extends AbstractN4JSDeclarativeValidator {
+	 * @param duplicateImportDeclaration The import declaration of the duplicated import
+	 * @param eObjectToIssueCode A map to keep track of all added issues
+	 */
+	protected def addIssueDuplicateNamedImportDeclaration(NamedImportSpecifier specifier,
```
can be changed to private, now

mor-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSTopLevelElementValidator.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * A validator to issue error messages on top-level elements based on the {@link JavaScriptVariant} in use.
+ */
+public class N4JSTopLevelElementValidator extends AbstractN4JSDeclarativeValidator {
```
move the one check method inside this class to `N4IDLVersionValidator`, delete this class, and then rename `N4IDLVersionValidator` to `N4IDLValidator`.

mor-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/BaseJavaScriptVariantHelper.java:
```diff
@@ -608,10 +629,41 @@ public String variantMode(EObject eobj) {
+	@Override
+	public String getVariantName(EObject eobj) {
+		return get(fileExtensionCalculator.getXpectAwareFileExtension(eobj), VARIANT_NAME);
+	}
```
not sure if the distinction between `N4JSJavaScriptVariantHelper` and `BaseJavaScriptVariantHelper` and interface `JavaScriptVariantHelper` still makes sense after N4JSX and N4IDL have been merged into N4JS.

Please double check this and discuss with interested parties.

lbeurerkellner replied:
I will discuss this with interested parties and create a follow-up issues.

mor-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/N4JSJavaScriptVariantHelper.java:
```diff
@@ -26,150 +27,237 @@
 	private final static String EXT_N4JS = N4JSGlobals.N4JS_FILE_EXTENSION;
 	private final static String EXT_N4JSD = N4JSGlobals.N4JSD_FILE_EXTENSION;
+	private final static String EXT_N4IDL = N4IDLGlobals.N4IDL_FILE_EXTENSION;
+	private final static String EXT_IDL = N4IDLGlobals.IDL_FILE_EXTENSION;
```
I'd remove all support for file extension ".idl" (but maybe discuss with Jens, first)

mor-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/GenerateN4JS.mwe2:
```diff
@@ -48,7 +48,7 @@ Workflow {
 		language = XtextGeneratorLanguage {
 			name = "org.eclipse.n4js.N4JS"
-			fileExtensions = "n4js,n4jsx,js,jsx,n4jsd"
+			fileExtensions = "n4js,n4jsx,js,jsx,n4jsd,n4idl,idl"
```
again, remove "idl"

mor-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/N4JSRuntimeModule.java:
```diff
@@ -544,7 +547,7 @@ public void configureFormatterPreferences(Binder binder) {
 	 */
 	public Class<? extends VersionResolver> bindVersionResolver() {
-		return N4JSVersionResolver.class;
+		return N4IDLVersionResolver.class;
```
the entire notion of `VersionResolver` could be removed (as a simplification)

mor-n4 commented on builds/org.eclipse.n4js.product.build/N4JS__IDE.launch:
```diff
@@ -49,8 +49,6 @@
 <setEntry value="org.eclipse.n4js.jsdoc2spec.sdk:default"/>
 <setEntry value="org.eclipse.n4js.lang.sdk:default"/>
-<setEntry value="org.eclipse.n4js.n4jsx.sdk:default"/>
-<setEntry value="org.eclipse.n4js.n4jsx.xpect.sdk:default"/>
```
why remove N4JSX from the launch configuration? Instead, I'd expect N4IDL to be added.

lbeurerkellner replied:
This is an automatic change that is caused by editing the launch configuration. These entries were removed by the UI, since the features "org.eclipse.n4js.n4jsx.sdk" and "org.eclipse.n4js.n4jsx.xpect.sdk" do not exist anymore.

mor-n4 commented on testhelpers/org.eclipse.n4js.tests.helper/src/org/eclipse/n4js/n4idl/tests/helper/N4IDLParseHelper.java:
```diff
@@ -0,0 +1,58 @@
+/**
+ * Replacement for ParseHelper in order to define mode in which file is parsed, also see {@link N4IDLParseHelper}.
+ */
+public class N4IDLParseHelper extends ParseHelper<Script> {
```
does it make sense to keep this as a separate class? I'd just add the #parseN4IDL() method to the ordinary `N4JSParseHelper`.

lbeurerkellner replied:
To me this was a logical decision, as it follows the pattern of `N4JSXParseHelper`, which is a separate class of similar nature.

mor-n4 commented on plugins/org.eclipse.n4js.transpiler/src/org/eclipse/n4js/transpiler/print/LineColTrackingAppendable.java:
```diff
@@ -51,6 +51,9 @@ public Appendable append(char c) throws IOException {
 	public Appendable append(CharSequence csq) throws IOException {
+		if (null == csq) {
+			return this;
+		}
```
this fail-safe behavior will just make bugs in the transpiler (or other code) be overlooked and more difficult to debug.
Please remove or throw an exception with a helpful message.

mor-n4 commented on plugins/org.eclipse.n4js.transpiler/src/org/eclipse/n4js/transpiler/AbstractTranspiler.java:
```diff
@@ -72,7 +72,7 @@
 	 * executed for the given initial transpiler state in the order they should be executed.
 	 */
-	protected abstract Transformation[] computeTransformationsToBeExecuted(TranspilerState state);
+	protected abstract Iterable<Transformation> computeTransformationsToBeExecuted(TranspilerState state);
```
why this change?

mor-n4 commented on plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ModuleWrappingTransformation.xtend:
```diff
@@ -740,4 +761,51 @@ class ModuleWrappingTransformation extends Transformation {
 		return assignmentStmnt;
 	}
+
+	protected def ParameterizedCallExpression createExportExpression(SymbolTableEntry entry, Expression expression) {
```
since this is N4IDL specific and we keep the separate N4IDL transpiler (for now), it should be moved to the N4IDL transpiler

mor-n4 commented on plugins/org.eclipse.n4js.ts/src/org/eclipse/n4js/ts/TypeExpressions.xtext:
```diff
@@ -152,12 +152,22 @@ ArrayTypeRef returns ParameterizedTypeRef:
+	(definedTypingStrategy=TypingStrategyUseSiteOperator declaredType=[Type|TypeReferenceName]
+	| {VersionedParameterizedTypeRefStructural} definedTypingStrategy=TypingStrategyUseSiteOperator
+	declaredType=[Type|TypeReferenceName] VersionRequest)
+	(-> TypeArguments)?
```
improve formatting to make more readable

mor-n4 commented:
general remark:
have you tested this with the headless builder (i.e. compiling n4ild files with n4jsc.jar from command line)? Or do we already have tests for this?

lbeurerkellner commented:
As discussed I made the following changes:

* I finished the work on the conceptual change of how imports work (see initial PR description)
* Make N4IDL-specific transpiler stages, if possible, exclusive to the N4IDLTranspiler
* Simplify class hierarchy of N4IDL-related scope providers
* Addressed other minor comments
* Removed bogus-support for 'idl' file extension
* Merged master

Builds are running.
